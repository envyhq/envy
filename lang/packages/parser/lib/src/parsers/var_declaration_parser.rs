use crate::{
    abstract_syntax_tree::{AbstractSyntaxNode, DeclarationNode, Leaf},
    parsers::attribute_block_parser::AttributeBlockParser,
    vars::{PartialVarDeclarationNode, VarDeclarationNode},
    Parser,
};
use nv_lexer::{
    tokens::{LexerSymbol, LexerToken},
    LexerKeyword, LexerTokenKind,
};
use std::sync::{Arc, Weak};

pub struct VarDeclarationParser;

impl Parser<Option<AbstractSyntaxNode>> for VarDeclarationParser {
    fn parse(
        tokens: &[LexerToken],
        parent: Weak<AbstractSyntaxNode>,
    ) -> (usize, Option<AbstractSyntaxNode>) {
        let mut buffer = vec![];
        let mut tokens_iter = tokens.iter().enumerate();

        let mut ast_fragment = None;

        let mut processed_count = 0;

        let mut partial_declaration = PartialVarDeclarationNode {
            parent: parent.clone(),
            identifier: None,
            type_value: None,
            modifier: None,
            attributes: vec![],
        };

        while let Some((index, token)) = tokens_iter.next() {
            processed_count += 1;

            match &token.kind {
                LexerTokenKind::Identifier(identifier) => {
                    partial_declaration.identifier =
                        Some(Leaf::new(identifier.clone(), token.range.clone()));

                    continue;
                }
                LexerTokenKind::Type(type_value) => {
                    partial_declaration.type_value =
                        Some(Leaf::new(type_value.clone(), token.range.clone()));

                    continue;
                }
                LexerTokenKind::Keyword(LexerKeyword::VarModifierKeyword(modifier)) => {
                    partial_declaration.modifier =
                        Some(Leaf::new(modifier.clone(), token.range.clone()));

                    continue;
                }
                LexerTokenKind::Symbol(LexerSymbol::BlockOpenCurly) => {
                    let (count, parsed_block) =
                        AttributeBlockParser::parse(&tokens[index..], parent.clone());

                    // -1 because we dont want to double count the block open curly
                    let count = count - 1;

                    partial_declaration.attributes = parsed_block
                        .iter()
                        .map(|attribute| attribute.clone().into())
                        .collect();

                    processed_count += count;
                    if count > 0 {
                        tokens_iter.nth(count - 1);
                    }

                    let declaration: Result<VarDeclarationNode, _> =
                        partial_declaration.clone().try_into();

                    if declaration.is_ok() {
                        ast_fragment = Some(AbstractSyntaxNode::Declaration(
                            DeclarationNode::VarDeclaration(Arc::new(declaration.unwrap())),
                        ));
                    }
                }
                LexerTokenKind::Symbol(LexerSymbol::Newline) => {
                    let declaration: Result<VarDeclarationNode, _> =
                        partial_declaration.clone().try_into();

                    if declaration.is_ok() {
                        ast_fragment = Some(AbstractSyntaxNode::Declaration(
                            DeclarationNode::VarDeclaration(Arc::new(declaration.unwrap())),
                        ));
                    }
                }
                _ => {
                    buffer.push(token);
                    continue;
                }
            };

            if ast_fragment.is_some() {
                return (processed_count, ast_fragment);
            }
        }

        let declaration: Result<VarDeclarationNode, _> = partial_declaration.clone().try_into();

        if declaration.is_ok() {
            ast_fragment = Some(AbstractSyntaxNode::Declaration(
                DeclarationNode::VarDeclaration(Arc::new(declaration.unwrap())),
            ));

            return (processed_count, ast_fragment);
        }

        (processed_count, ast_fragment)
    }
}

#[cfg(test)]
mod tests {
    use crate::parsers::var_declaration_parser::{Parser, VarDeclarationParser};
    use nv_lexer::{Lexer, TokenPosition, VarDeclarationLexer};
    use nv_unit_testing::str_to_graphemes;
    use std::sync::Weak;

    #[test]
    fn parses_var_nodes() {
        // We dont have "var" keyword here, its handled by the source file parser
        let input = str_to_graphemes("my_cool_var: str");

        let start_line = 10;
        let start_column = 5;
        let mut lexer =
            VarDeclarationLexer::new(&input, TokenPosition::new(start_line, start_column));
        lexer.lex();

        let parent = Weak::new();
        let (count, ast) = VarDeclarationParser::parse(&lexer.tokens, parent);

        assert_eq!(count, 3);

        insta::assert_yaml_snapshot!(ast);
    }
}
