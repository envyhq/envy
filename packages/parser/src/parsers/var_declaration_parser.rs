use crate::{
    abstract_syntax_tree::{AbstractSyntaxNode, DeclarationNode, Leaf},
    parsers::attribute_block_parser::AttributeBlockParser,
    vars::{PartialVarDeclarationNode, VarDeclarationNode},
    Parser,
};
use nv_lexer::{
    tokens::{LexerSymbol, LexerToken},
    LexerKeyword, LexerTokenKind,
};
use std::sync::Weak;

pub struct VarDeclarationParser;

impl Parser<Option<AbstractSyntaxNode>> for VarDeclarationParser {
    fn parse(
        tokens: &Vec<LexerToken>,
        parent: Weak<AbstractSyntaxNode>,
    ) -> (usize, Option<AbstractSyntaxNode>) {
        let mut buffer = vec![];
        let mut tokens_iter = tokens.iter().enumerate();

        let mut ast_fragment = None;

        let mut processed_count = 0;

        let mut partial_declaration = PartialVarDeclarationNode {
            parent: parent.clone(),
            identifier: None,
            type_value: None,
            modifier: None,
            attributes: vec![],
        };

        while let Some((index, token)) = tokens_iter.next() {
            let token = token.to_owned();

            processed_count += 1;

            let sub_tokens = &tokens[index..].to_vec();
            let sub_tokens = sub_tokens.to_vec();

            match token.kind {
                LexerTokenKind::Identifier(identifier) => {
                    partial_declaration.identifier = Some(Leaf::new(identifier, token.range));

                    continue;
                }
                LexerTokenKind::Type(type_value) => {
                    partial_declaration.type_value = Some(Leaf::new(type_value, token.range));

                    continue;
                }
                LexerTokenKind::Keyword(LexerKeyword::VarModifierKeyword(modifier)) => {
                    partial_declaration.modifier = Some(Leaf::new(modifier, token.range));

                    continue;
                }
                LexerTokenKind::Symbol(LexerSymbol::BlockOpenCurly) => {
                    let (count, parsed_block) =
                        AttributeBlockParser::parse(&sub_tokens, parent.clone());

                    // -1 because we dont want to double count the block open curly
                    let count = count - 1;

                    partial_declaration.attributes = parsed_block
                        .iter()
                        .map(|attribute| attribute.clone().into())
                        .collect();

                    processed_count += count;
                    if count > 0 {
                        tokens_iter.nth(count - 1);
                    }

                    let declaration: Result<VarDeclarationNode, _> =
                        partial_declaration.clone().try_into();

                    if declaration.is_ok() {
                        ast_fragment = Some(AbstractSyntaxNode::Declaration(
                            DeclarationNode::VarDeclaration(declaration.unwrap()),
                        ));
                    }
                }
                LexerTokenKind::Symbol(LexerSymbol::Newline) => {
                    let declaration: Result<VarDeclarationNode, _> =
                        partial_declaration.clone().try_into();

                    if declaration.is_ok() {
                        ast_fragment = Some(AbstractSyntaxNode::Declaration(
                            DeclarationNode::VarDeclaration(declaration.unwrap()),
                        ));
                    }
                }
                _ => {
                    buffer.push(token);
                    continue;
                }
            };

            if ast_fragment.is_some() {
                return (processed_count, ast_fragment);
            }
        }

        let declaration: Result<VarDeclarationNode, _> = partial_declaration.clone().try_into();

        if declaration.is_ok() {
            ast_fragment = Some(AbstractSyntaxNode::Declaration(
                DeclarationNode::VarDeclaration(declaration.unwrap()),
            ));

            return (processed_count, ast_fragment);
        }

        (processed_count, ast_fragment)
    }
}
