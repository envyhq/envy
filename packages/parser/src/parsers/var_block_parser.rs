use crate::{abstract_syntax_tree::AbstractSyntaxNode, parser::ParserResult, Parser};
use nv_lexer::{
    tokens::LexerDeclarationKeyword, LexerKeyword, LexerTokenKind, LexerVarModifierKeyword,
};

use super::var_declaration_parser::VarDeclarationParser;

pub struct VarBlockParser {
    pub ast_block: Vec<AbstractSyntaxNode>,
    pub tokens: Vec<LexerTokenKind>,

    buffer: Vec<LexerTokenKind>,
}

impl Parser for VarBlockParser {
    fn parse(&mut self) -> usize {
        let bound_tokens = self.tokens.clone();
        let mut tokens = bound_tokens.iter().enumerate();

        self.buffer.clear();

        let mut processed_count = 0;

        while let Some((index, token)) = tokens.next() {
            processed_count += 1;
            let sub_tokens = &bound_tokens[index..].to_vec();
            let sub_tokens = sub_tokens.to_vec();

            let result = match token {
                LexerTokenKind::Keyword(LexerKeyword::VarModifierKeyword(
                    LexerVarModifierKeyword::Pub,
                ))
                | LexerTokenKind::Keyword(LexerKeyword::DeclarationKeyword(
                    LexerDeclarationKeyword::Var,
                )) => {
                    let mut parser = VarDeclarationParser::new(sub_tokens.clone());
                    // -1 to avoid double counting the leading token (var or pub)
                    let count = parser.parse() - 1;

                    if let Some(ast_fragment) = parser.ast_fragment.clone() {
                        self.ast_block.push(ast_fragment);
                    }

                    ParserResult {
                        processed_count: count,
                        ast_fragment: parser.ast_fragment,
                    }
                }
                _ => {
                    continue;
                }
            };

            processed_count += result.processed_count;
            if result.processed_count > 0 {
                tokens.nth(result.processed_count - 1);
            }
        }

        processed_count
    }
}

impl VarBlockParser {
    pub fn new(tokens: Vec<LexerTokenKind>) -> Self {
        Self {
            ast_block: vec![],
            tokens,
            buffer: vec![],
        }
    }
}
