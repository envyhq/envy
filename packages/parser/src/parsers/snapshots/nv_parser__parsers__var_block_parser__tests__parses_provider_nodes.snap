---
source: packages/parser/src/parsers/var_block_parser.rs
expression: ast
---
[
    Declaration(
        VarDeclaration(
            VarDeclarationNode {
                identifier: Leaf {
                    value: "my_cool_var",
                    range: TokenRange {
                        from: TokenPosition {
                            line: 14,
                            column: 9,
                        },
                        to: TokenPosition {
                            line: 14,
                            column: 19,
                        },
                    },
                },
                type_value: Leaf {
                    value: String,
                    range: TokenRange {
                        from: TokenPosition {
                            line: 14,
                            column: 22,
                        },
                        to: TokenPosition {
                            line: 14,
                            column: 24,
                        },
                    },
                },
                modifier: None,
                attributes: [],
                parent: (Weak),
            },
        ),
    ),
    Declaration(
        VarDeclaration(
            VarDeclarationNode {
                identifier: Leaf {
                    value: "my_other_var",
                    range: TokenRange {
                        from: TokenPosition {
                            line: 15,
                            column: 13,
                        },
                        to: TokenPosition {
                            line: 15,
                            column: 24,
                        },
                    },
                },
                type_value: Leaf {
                    value: String,
                    range: TokenRange {
                        from: TokenPosition {
                            line: 15,
                            column: 27,
                        },
                        to: TokenPosition {
                            line: 15,
                            column: 29,
                        },
                    },
                },
                modifier: Some(
                    Leaf {
                        value: Pub,
                        range: TokenRange {
                            from: TokenPosition {
                                line: 15,
                                column: 5,
                            },
                            to: TokenPosition {
                                line: 15,
                                column: 7,
                            },
                        },
                    },
                ),
                attributes: [],
                parent: (Weak),
            },
        ),
    ),
]
